src/SDK/Trace/SpanProcessor/BatchSpanProcessor.php

```diff
@@ -93,7 +93,9 @@ public function onEnd(ReadableSpanInterface $span): void
         }
 
         if ($this->bufferReachedExportLimit() || $this->enoughTimeHasPassed()) {
-            $this->forceFlush();
+            $this->exporter->export($this->queue);
+            $this->queue = [];
```

Comment:
I have created a try block inside which export would take place. If somehow exporter suspends execution and fails , finally block would clear the queue. Will this help resolving the issue??

Reply:
No, I meant suspend as in `Fiber::suspend()`. Something like the following would resolve it:
```php
$batch = $this->queue;
$this->queue = [];
$this->stopwatch->reset();

try {
    $this->exporter->export($batch);
} finally {
    $this->queueSize -= count($batch);
}
```

---

The spec mentions that "[Export()](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#exportbatch) will never be called concurrently for the same exporter instance", but I don't see a way for us to follow the spec unless we queue execution in a single fiber. This would mean even more blocking for the [Span::end()](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#end) call that triggered the initial export (worst case blocking forever) ("Expect this operation to be called in the "hot path" of production applications. It needs to be designed to complete fast, if not immediately.").